include/pybind11/stl_bind.h

```diff
@@ -326,6 +326,48 @@ template <typename Vector, typename Class_> auto vector_if_insertion_operator(Cl
     );
 }
 
+// Provide the buffer interface for vectors if we have data() and we have a format for it
+// GCC seems to have "void std::vector<bool>::data()" - doing SFINAE on the existence of data() is insufficient, we need to check it doesn't return void
+template <typename Vector, typename = void>
+struct vector_has_data_and_format : std::false_type {};
+template <typename Vector>
+struct vector_has_data_and_format<Vector, enable_if_t<std::is_same<decltype(py::format_descriptor<typename Vector::value_type>::format(), std::declval<Vector>().data()), typename Vector::value_type*>::value>> : std::true_type {};
+
+// Construct a vector wrapping class, with the buffer interface if possible
+template <typename Vector, typename Class_, typename... Args>
+enable_if_t<vector_has_data_and_format<Vector>::value, Class_>
+vector_buffer(Args&&... args) {
+    using T = typename Vector::value_type;
+
+    try {
+        // numpy.h declares this for arbitrary types, but it may raise an exception if PYBIND11_NUMPY_DTYPE hasn't been called
+        py::format_descriptor<T>::format();
+    } catch (std::runtime_error&) {
+        return Class_(std::forward<Args>(args)...);
+    }
+
+    Class_ cl = Class_(std::forward<Args>(args)..., py::buffer_protocol());
+
+    cl.def_buffer([](Vector& v) -> py::buffer_info {
+        return py::buffer_info(v.data(), sizeof(T), py::format_descriptor<T>::format(), 1, {v.size()}, {sizeof(T)});
+    });
+
+    cl.def("__init__", [](Vector& vec, py::buffer buf) {
+        auto info = buf.request();
+        if (info.ndim != 1)
+            throw pybind11::type_error("Only 1D buffers can be copied to a vector");
+        if (info.strides[0] != sizeof(T))
+            throw pybind11::type_error("Item size mismatch (Python: " + std::to_string(info.strides[0]) + " C++: " + std::to_string(sizeof(T)) + ")");
```

Comment:
I don't think this `strides[0] == sizeof(T)` requirement is desirable: it should be easy enough to support arbitrary strides, which let slices (e.g. single rows/columns of numpy matrices) work as vector buffer inputs.  Something like this (untested) as a replacement for the Vector constructor ought to work:

```C++
new (&vec) Vector();
vec.reserve(info.shape[0]);
for (void *ptr = info.ptr, void *end = info.ptr + info.strides[0] * info.shape[0]; ptr != end; ptr += info.strides[0])
    vec.push_back(*static_cast<T *>(ptr));
```

(A small test to see that this really does work would be good as well).